appserver/ejb/ejb-http-remoting/client/src/main/java/fish/payara/ejb/http/client/RemoteEJBContext.java

```diff
@@ -133,130 +132,151 @@ public Object lookup(String name) throws NamingException {
                 }
             }
             // Get client build with all optional config applied
-            ClientBuilder clientBuilder = getClientBuilder();
-            
-            // For the lookup do a call to the remote server first to obtain
-            // the remote business interface class name given a JNDI lookup name.
-            // The JNDI lookup name normally does not give us this interface name.
-            // This also allows us to check the JNDI name indeed resolves before
-            // we create a proxy and return it here.
-            
-            Map<String, Object> payload = new HashMap<>();
-            payload.put("lookup", name);
-            
             URI remotePayaraURI = new URL(url).toURI();
-            
-            String className = 
-                clientBuilder
-                    .build()
-                    .target(remotePayaraURI)
-                    .path("ejb")
-                    .path("lookup")
-                    .request()
-                    .post(Entity.entity(payload, APPLICATION_JSON))
-                    .readEntity(String.class);
-            
-            // After we have obtained the class name of the remote interface, generate
-            // a proxy based on it.
-            
-            return EjbHttpProxyFactory.newProxy(
-                    Class.forName(className), 
-                    clientBuilder.build()
-                                 .target(remotePayaraURI)
-                                 .path("ejb")
-                                 .path("invoke"), 
-                    name,
-                    new HashMap<String, Object>(environment)
-                    );
-            
+
+            return lookup(remotePayaraURI).lookup(name);
+        } catch (NamingException ne) {
+            throw ne;
         } catch (Exception e) {
             throw newNamingException(name, e);
         }
     }
-    
+
+
+
+    /**
+     * Lazily initialize lookup strategy fit for required serialization and server-side supported protocol.
+     * @return
+     */
+    private synchronized Lookup lookup(URI root) throws NamingException {
+        if (lookup == null) {
+            lookup = determineLookup(root);
+        }
+        return lookup;
+    }
+
+    /**
+     * Determine supported protocol on server side and construct corresponding Lookup instance.
+     */
+    private Lookup determineLookup(URI root) throws NamingException {
+        ClientBuilder clientBuilder = null;
+        try {
+            clientBuilder = getClientBuilder();
+            Client client = clientBuilder.build();
+
+            LookupDiscoveryServiceImpl lookupDiscovery = new LookupDiscoveryServiceImpl();
+            LookupDiscoveryResponse discovery = lookupDiscovery.discover(client, root);
+
+            Integer version = getVersion(discovery);
+            if (discovery.isV1target() && (version == null || version.intValue() == 1)) {
```

Comment:
Haven't you just done this check in the `getVersion` method?
`discovery.isV1target() && (version == null || version.intValue() == 1)`
vs.
`version == 1 && discovery.isV1target()`

Reply:
Not exactly. The version selection is a bit delicate. The goal is to use the version specified if specified *and* available otherwise fallback on the newest available version. This part is the fallback while the `getVersion` method returns the selected version if specified and available.